tutorials/use-cloud-pubsub-cloud-storage-app-engine/index.md

```diff
@@ -0,0 +1,1357 @@
+1. If you want collaborators on your photo album, [add IAM Permissions](https://www.google.com/url?sa=D&q=https%3A%2F%2Fcloud.google.com%2Fstorage%2Fdocs%2Faccess-control%2Fusing-iam-permissions%23bucket-add) to your bucket. Select`Storage Object Admin` as the role for each collaborator you want to add.
+1. Change the photos bucket permissions to make it publicly readable so that the photos may be viewed on your website. To do this, you'll need to [make your bucket data public](https://www.google.com/url?sa=D&q=https%3A%2F%2Fcloud.google.com%2Fstorage%2Fdocs%2Faccess-control%2Fmaking-data-public%23buckets).
+
+1. Create another bucket with `Multi-Regional` or `Regional` storage. This bucket is for storing the thumbnails of the photos in your shared photo album.
+1. [Create a new topic](https://www.google.com/url?sa=D&q=https%3A%2F%2Fcloud.google.com%2Fpubsub%2Fdocs%2Fquickstart-console%23create_a_topic) with the same name as your photos bucket.
+1. Click on the three-dots icon for your photo album topic and click on `New subscription`. Change the `Delivery Type` to `Push into an endpoint url`. This is the url that receives your Cloud Pub/Sub messages. For a url, use the following, replacing [PROJECT-ID] with the name of your project: `http://[PROJECT ID].appspot.com/_ah/push-handlers/receive_message`.
+1. Configure Cloud Pub/Sub notifications for your photos bucket by using the command line to run:
+
+    ```sh
+    gsutil notification create -f json gs://[PHOTO BUCKET NAME]
+    ```
+
+## Basic application layout
+
+If you do not feel like coding the entire application from scratch, feel free to clone the git repository with a default application by running:
+
+  ```sh
+  git clone https://github.com/GChien44/tutorial-v2.git
+  ```
+
+Note that if you choose this option, you still need to make a `lib` directory and run the `install` command. These tasks are outlined in steps one and three of the **Libraries and `app.yaml`** section. In addition to these tasks, some constants in the `main.py` file still need to be changed to suit your GCS bucket names. Look for the constants THUMBNAIL_BUCKET, PHOTO_BUCKET, NUM_NOTIFICATIONS_TO_DISPLAY, and MAX_LABELS immediately after the imports at the top of the file.
+
+The rest of this tutorial assumes that you did not clone the above git repository, and are building the application from scratch.
+
+### Libraries and `app.yaml`
+
+The external library and `app.yaml` files are necessary for configuring your App Engine application and importing the required libraries.
+
+1. Choose a directory to house your project. From this point forward, this will be referred to as the host directory. Inside your host directory, create a new directory called `lib` for the storage of external libraries.
+1. In your host directory, create the file `requirements.txt` and copy in the following text:
+
+    ```txt
+    jinja2
```

Comment:
you should specify specific versions of these packages to avoid new releases breaking this tutorial.